### DISCUSSION TITLE
Hydrogen on Windows 10 - no MIDI devices appear

### DISCUSSION THREAD

**Discussion by DrJanNielsen** [hydrogen-music/hydrogen#1047] [Hydrogen on Windows]
Apologies if this is a misunderstanding but I'm stumped.

I have installed Hydrogen 1.0.1 on Windows 10; installation went fine and hydrogen starts and works as software only system.

But I want to experiment with using hydrogen to output MIDI to control my BOSS DR-880 drum synth. However, When I go to select my MIDI ports I only have "None" as options for both Input and Output.

I have tried to run PmDefaults (which does find all my MIDI devices (both virtual and physical) but nothing changes in Hydrogen.

Am I missing something to pick up a configuration from somewhere? If there is an answer I promise to provide a documentation patch.

**Comment by catopede** [comment]
I have a possibly related problem. Although Hydrogen defaults to "no MIDI input device", it successfully detects and lists my Yamaha YS as an option, However, when I try to re-start the driver specifying the Yamaha, Hydrogen crashes.

**Reply by cme** [reply]
Is that, by any chance, on 64-bit Windows? If so, that's a known PortMidi issue with the version packaged in MSYS2 which is where Hydrogen gets its dependencies from.

It's been fixed in MSYS2 so we should push out a Hydrogen release soon to pick up this and some other fixes.

**Comment by catopede** [comment]
Yes, it is 64 bit Windows... Thanks for the info, good to hear there's a
solution on the way!

Julian

**Comment by NoAliby** [comment]
I had the same issue. When selecting the Midi input device it want's to restart and then forgets the midi device selection. I am also on Windows 64bit. I uninstalled the latest Hydrogen 1.1.0 and reverted back to Hydrogen 1.1.0 beta1 (or 1.02) and now the Midi input device works fine.

**Reply by cme** [reply]
This sounds like it's more likely to be #1375 than a MIDI-specific issue?

**Comment by cme** [comment]
Hi @DrJanNielsen , @catopede and @NoAliby ,

Would you be able to try these preview builds, please? 

  * [Windows 32-bit](https://ci.appveyor.com/api/buildjobs/ht2malp892wapq2d/artifacts/Hydrogen-1.1.1-win32.exe)
  * [Windows 64-bit](https://ci.appveyor.com/api/buildjobs/g3xqkokj2a44a8k8/artifacts/Hydrogen-1.1.1-win64.exe)

These should address MIDI issues as well as preferences dialog issues. All being well these will be representative of what will be the 1.1.1 release :)

**Comment by DrJanNielsen** [comment]
I’ll get set up tomorrow and give it a whirl

Jan

**Reply by cme** [reply]
Awesome! If the issue is still there, would you be able to help a little with debugging by running an augmented debug version of Hydrogen and capturing logs? Hardware-related issues are sadly something that are difficult to debug since we naturally don't have access to the same hardware! :D Thanks!

**Comment by DrJanNielsen** [comment]
Sure

Jan

**Comment by DrJanNielsen** [comment]
All,

I just had a moment to download the test version of Hydrogen re this issue.

The good news is that hydrogen now sees all the various midi interfaces on
the PC (same PC and setup as my original post). I have also managed to
capture (record) midi events generated on my Boss DR-880 in Hydrogen. Not
quite figured out how to play events out but I’m certain it will work once
I get the mappings set up correctly (and I get a bit more familiar with
Hydrogen and its handling of MIDI).

Under any circumstances, great catch and and thanks for looking into this

Cheers

Jan

**Reply by cme** [reply]
Fantastic! Thanks very much, and fingers crossed for the next step :D

**Comment by NoAliby** [comment]
The 64bit seems to work OK now. One thing I noticed is that before when it prompted that it needed to restart the driver Hydrogen would close. Now it prompts you, but the program stays open. That is a plus.

However, I am getting an error now when trying to delete a pattern "Could not save pattern to temporary directory"
Will revert to version 1.1.0 32 bit

**Reply by cme** [reply]
That's... interesting. I don't think there are any changes between 1.1.0 and now that would cause that behaviour.

Would you be able to check what the temporary directory in use is? You can get this from `Debug` -> `Show filesystem info` and the `Temporary directory` should show something like `C:/Users/username/AppData/Local/Temp/hydrogen`, and check that exists by hitting 'Open'?

Thanks for for your help!

**Reply by NoAliby** [reply]
Sorry that was actually an issue if the song had not been saved or inn some cases if the part had a strange character in it. Seems 1.1.1 is fine

**Reply by NoAliby** [reply]
FYI, I still get this event. (v1.1.1 64bit)
"Could not save pattern to temporary directory"
So far I can confirm it "can" occur when DUPLICATING or DELETING a pattern from the Song section.

I think it must be something to do with a caching or memory issue. If I CLOSE Hydrogen and open it again, open the song file, then the issue is gone and I can usually carry on programming my Drum Song.

_Let me know if there is anything else you need to help troubleshoot it._

**Comment by cme** [comment]
If you can reproduce this issue, could you please enable debug logging ( Debug &rArr; Log level &rArr; Debug) and capture the log file (c:\Users\username\.hydrogen\hydrogen.log) when in happens? That will show which filenames it's trying to use and the location it's trying to access.

**Reply by cme** [reply]
Aha! It looks like we're not escaping pattern names and using them unaltered as filenames for temporary files. That will cause problems if anything in the pattern name isn't a valid Windows filename.

**Reply by cme** [reply]
Have created #1425 for this.

**Comment by NoAliby** [comment]
@cme, is there a more recent build?
I have 1.1.1-'b39fcf77' [Apr 21, 2022]. I can't remember if it was from you or @theGreatWhiteShark. I don't know where can I find builds you guys are compiling.

While this pre-release version had fixed some issues I was having there are some nagging items, I would love if they went "away":
1. If I move notes from one instrument to another they are silent - same happens is I duplicate a track.
2. The undo does not include a change in bar length of loop. So if you reduce the length of a loop you loose what you had programmed and hopefully you saved just before you did or you lose the work.

If there is a newer build I am hoping it has those aspects sorted.

**Reply by theGreatWhiteShark** [reply]
Hey @NoAliby ,

> While this pre-release version had fixed some issues I was having there are some nagging items, I would love if they went "away"

I think this should be fixed by now. I prepared you the latest pre-release so you can check yourself and report back ;)
https://ci.appveyor.com/api/buildjobs/7e1jadswp5qjmegv/artifacts/Hydrogen-1.1.1-win64.exe
https://ci.appveyor.com/api/buildjobs/vybayrn754lkc7sn/artifacts/Hydrogen-1.1.1-win32.exe

**Reply by cme** [reply]
> I think this should be fixed by now. I prepared you the latest pre-release so you can check yourself and report back ;) https://ci.appveyor.com/api/buildjobs/7e1jadswp5qjmegv/artifacts/Hydrogen-1.1.1-win64.exe https://ci.appveyor.com/api/buildjobs/vybayrn754lkc7sn/artifacts/Hydrogen-1.1.1-win32.exe

I meant to make a 1.1.2 maintenance release earlier in the year, and should probably do that before making a 1.2 beta happen. @theGreatWhiteShark is that build from `releases/1.1`? And  @NoAliby does that solve your issues?

**Reply by theGreatWhiteShark** [reply]
> @theGreatWhiteShark is that build from releases/1.1

Nope. It's the latest version of `master` d3d8e951002798505b2511cb9b41b7eb5829bf92 . (We should use a dedicated version for the master branch, like `1.1.1-9000` or whatever semantic versioning suggest in such cases, to tell its builds apart from those of the latest release)

**Reply by NoAliby** [reply]
I can confirm that the undo is now working.
I will get back to you if I find the move/silence issue occurs again.
Thanks

**Comment by nunosga** [comment]
Hi,
I'm trying to use a midi keyboard (Korg Microkey) to play / record patterns in Hydrogen, but i can't get Hydrogen to work with the keyboard. (the Korg works fine with the studio one DAW).
I'm not a programmer (only made a few batch (MS-Dos) code lines in the day).  What should i do?

**Comment by BigD71** [comment]
Dear Sirs,
i have the same problems with Windows 10 64Bit. My MIDI-Device is a Donner DEP-20 Piano. My Piano works well on other DAW's but would not be recognized by Hydrogen. I've tried every available 64 and 32 Bit versions, but no luck at all. SEQ24 uses also PortMIDI and is working well.
Please give me an advise.

Kind regards
